wallet-new/src/Cardano/Wallet/Kernel/BListener.hs

```diff
@@ -49,21 +61,117 @@ prefilterBlock' pw b = do
+        % " }"
+        )
+        hh
+
 -- | Notify all the wallets in the PassiveWallet of a new block
+--
+-- NOTE: Multiple concurrent or parallel calls to 'applyBlock' are not allowed.
+-- Without this constraint, two concurrent backfill operations could conflict
+-- with each other and cause both to fail.
+-- The serialization of calls to 'applyBlock' is handled by the wallet worker,
+-- which should carry the sole responsibility for applying blocks to a wallet.
 applyBlock :: PassiveWallet
            -> ResolvedBlock
-           -> IO (Either ApplyBlockFailed ())
+           -> IO (Either BackfillFailed ())
 applyBlock pw@PassiveWallet{..} b = do
     k <- Node.getSecurityParameter _walletNode
-    ((ctxt, blocksByAccount), metas) <- prefilterBlock' pw b
-    -- apply block to all Accounts in all Wallets
-    mConfirmed <- update' _wallets $ ApplyBlock k ctxt blocksByAccount
-    case mConfirmed of
-      Left  err       -> return $ Left err
-      Right confirmed -> do
-        modifyMVar_ _walletSubmission $ return . Submission.remPending confirmed
-        mapM_ (putTxMeta _walletMeta) metas
-        return $ Right ()
+    runExceptT (applyOneBlock k b) >>= \case
+        Right () -> return (Right ())
+        Left  (ApplyBlockNotSuccessor curCtx cpCtx) -> runExceptT $ do
+          -- If we could not apply this block, it may be because the wallet has
+          -- fallen behind the node and is missing blocks. Try to find and apply
+          -- each of those blocks as well.
+          blocks <- backfill (Just curCtx) cpCtx []
+          for_ (getOldestFirst blocks) (withExceptT convertError . applyOneBlock k)
+
+  where
+
+      -- Try to apply a single block, failing if it does not fit onto the most recent checkpoint.
+      applyOneBlock :: Node.SecurityParameter -> ResolvedBlock -> ExceptT ApplyBlockFailed IO ()
+      applyOneBlock k b' = ExceptT $ do
+          ((ctxt, blocksByAccount), metas) <- prefilterBlock' pw b'
+          -- apply block to all Accounts in all Wallets
+          mConfirmed <- update' _wallets $ ApplyBlock k ctxt blocksByAccount
+          case mConfirmed of
+              Left  err -> return (Left err)
+              Right confirmed -> do
+                  modifyMVar_ _walletSubmission (return . Submission.remPending confirmed)
+                  mapM_ (putTxMeta _walletMeta) metas
+                  return $ Right ()
+
+      -- Interpret an ApplyBlockFailed error during backfilling as a BackfillFailed.
+      convertError :: ApplyBlockFailed -> BackfillFailed
+      convertError = \case
+          ApplyBlockNotSuccessor curCtx cpCtx -> SuccessorChanged curCtx cpCtx
+
+      -- Backfill blocks that were missing between the given block and the wallet's most recent
+      -- checkpoint. 'backfill src tgt' will apply all blocks in order, starting from the
+      -- successor of tgt until src is reached. 'Nothing' is used to represent the genesis block.
+      backfill :: Maybe BlockContext
```

Comment:
I guess minor misnomer here, since this is not doing any actual backfilling anymore.